feat: conform zeebe backup to contract changes

panagiotisgts opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
The scope of this PR is to conform the existing backup API to the changes required to handled scheduled backups. Also, the current API is only responsible for emitting `CheckpointType#MANUAL_BACKUP` records. 

- Make use of the `BackupDescriptor` to propagate checkpoint information down the backup API callstack
- Extend `BackupDescriptor` to also include the `checkpointId`
- `BackupDescriptor#checkpointId` is always initialized from the parent `BackupIdentifier#checkpointId` and is not part of the serialized output
- `BackupService` only relies on a complete `BackupDescriptor`
- Interpartition command send/receive includes `CheckpointType` as well
- `CheckpointCreateProcessor` conditionally takes a backup depending on the `CheckpointType`

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #40577

panagiotisgts committed:
feat: descriptor contains checkpointId; propagate descriptor down the stack

We also make sure that the descriptor.checkpointId is always derived
from the header backupidentifier id when deserializing

Use the BackupDescriptor to propagate information down the backup chain

panagiotisgts committed:
test: align backup store test with backup descriptor

panagiotisgts committed:
fix: backupservice utilize whole descriptor

panagiotisgts committed:
fix: BackupApiRequestHandler explicitly set type to manual backup

panagiotisgts committed:
feat: aling CheckpointCreateProcessor

Conform the processor with the new backup api.

Processor conditionally takes a backup based on the checkpoint type.

panagiotisgts committed:
feat: CheckpointListener propagates checkpoint type

panagiotisgts committed:
feat: interpartition messaging handles checkpoint type

panagiotisgts committed:
fix: checkpoint position is derived from wrapper record

panagiotisgts committed:
style: spotless apply

lenaschoenburg commented on zeebe/protocol/src/main/java/io/camunda/zeebe/protocol/record/value/management/CheckpointType.java:
```diff
@@ -16,7 +16,46 @@
+      default:
+        {
+          return null;
+        }
```
This is relevant in case we ever want to introduce a new checkpoint type 👍 

💭 I'm wondering if it's safe if just fall back to `MANUAL_BACKUP` then. At worst, we'd take a backup when we don't actually want one, right?

lenaschoenburg commented on zeebe/broker/src/main/java/io/camunda/zeebe/broker/transport/partitionapi/InterPartitionCommandReceiverImpl.java:
```diff
@@ -150,8 +157,9 @@ DecodedMessage decodeMessage(final byte[] message) {
       final var recordMetadata = new RecordMetadata();
       final var value = decodeCommand(messageDecoder, recordMetadata);
       decodeAuthInfo(messageDecoder, recordMetadata);
+      final var checkpointType = CheckpointType.valueOf(messageDecoder.checkpointType());
```
:x: Once we have set a new protocol version and `sinceVersion`, we'll need to check whether this field is available yet or not.

lenaschoenburg commented on zeebe/broker/src/main/resources/broker-protocol.xml:
```diff
@@ -20,6 +20,7 @@
     <field name="intent" id="2" type="uint8"/>
     <field name="recordKey" id="3" type="uint64" presence="optional"/>
     <field name="checkpointId" id="4" type="int64"/>
+    <field name="checkpointType" id="5" type="uint8"/>
```
Yes, you bump the version in the xml header and set `sinceVersion` on the field. For an example you can check out `SegmentDescriptor` in the journal schema and the generated deserializer code in `io.camunda.zeebe.journal.file.SegmentDescriptorDecoder#lastPosition`

panagiotisgts committed:
fix: bump sbe protocol version

Handle past protocol version empty checkpoint type as manual backup for compatibility

panagiotisgts committed:
fix: remove checkpointId from InProgressBackup

It's already part of the backup identifier

panagiotisgts committed:
Revert "test: align backup store test with backup descriptor"

This reverts commit 486e0b11f84e27f4c96a6f2d29d63fb29f82319c.

panagiotisgts committed:
Revert "feat: descriptor contains checkpointId; propagate descriptor down the stack"

This reverts commit bb8e530c

panagiotisgts committed:
fix: align services with externalized checkpointId

checkpointId is no longer part of the backup descriptor, align the
call stack to provide checkpointId externally

lenaschoenburg reviewed (approved):
Nicely done 🚀

panagiotisgts committed:
refactor: minor clean up
